### PR TITLE
Add API to get list of charsuites, special days, and image sources

### PR DESCRIPTION
**Pull request by cpeel**
This PR extends the API to allow users to get a list of character suites, special days, and image sources. This will be necessary for users to create projects (eventually) as part of [Task 2011](https://www.pgdp.net/c/tasks.php?action=show&task_id=2011).

The API intentionally does not use the exact same field names as those stored in the database. Instead, we try to present a more unified view to the outside world. Like other parts of the API, these only return data that the user can already see in the UI. This is most relevant to image sources which have scoped visibility -- some can only be viewed by PMs, or Image Source Managers / SAs. The PR includes update to the Swagger schema as well.

This code is live in the sandbox at https://www.pgdp.org/~cpeel/c.branch/api-proj-gets/ but there are no UI changes. Instead, testing can be done via `curl`:

```bash
# export the user's API key
# this one is for an unprivileged proofreader on TEST:
export API_KEY=review

# charsuites
curl -i -g -X GET "https://www.pgdp.org/~cpeel/c.branch/api-proj-gets/api/index.php?url=v1/projects/charsuites" \
    -H "accept: application/json" \
    -H "X-API-KEY: $API_KEY"

# special days
curl -i -g -X GET "https://www.pgdp.org/~cpeel/c.branch/api-proj-gets/api/index.php?url=v1/projects/specialdays" \
    -H "accept: application/json" \
    -H "X-API-KEY: $API_KEY"

# image sources
curl -i -g -X GET "https://www.pgdp.org/~cpeel/c.branch/api-proj-gets/api/index.php?url=v1/projects/imagesources" \
    -H "accept: application/json" \
    -H "X-API-KEY: $API_KEY"
```

All of these support an `enabled` parameter that, if present, will only return the entities that are enabled.
```bash
# only enabled image sources
curl -i -g -X GET "https://www.pgdp.org/~cpeel/c.branch/api-proj-gets/api/index.php?url=v1/projects/imagesources&enabled" \
    -H "accept: application/json" \
    -H "X-API-KEY: $API_KEY"
```

### CODE REVIEW

File: api/v1.inc [hunk 1]
```diff
@@ -25,6 +25,9 @@ $router->add_route("GET", "v1/projects/genres", "api_v1_projects_genres");
 $router->add_route("GET", "v1/projects/languages", "api_v1_projects_languages");
 $router->add_route("GET", "v1/projects/states", "api_v1_projects_states");
 $router->add_route("GET", "v1/projects/pagerounds", "api_v1_projects_pagerounds");
+$router->add_route("GET", "v1/projects/charsuites", "api_v1_projects_charsuites");
```
Comment (chrismiceli) [inline api/v1.inc:28]: I know our project ids cannot resemble words like "states", or "charsuites", but for routing purposes, does order matter? That way an ambiguous route like "v1/projects/states" that can match both won't ever hit the "v1/projects/:projectid" route with :projectid set to "states"? I guess if it is working, it doesn't matter.

File: api/v1_projects.inc [hunk 2]
```diff
@@ -316,3 +317,86 @@ function api_v1_projects_pagerounds($method, $data, $query_params)
...
+            continue;
+        }
+        $return_data[] = [
+            "id" => $special_day["spec_code"],
+            "name" => $special_day["display_name"],
+            "comment" => $special_day["comment"],
+            "url" => $special_day["info_url"],
+            "color" => $special_day["color"],
+            "symbol" => $special_day["symbol"],
+            "date_open" => sprintf("%02d-%02d", $special_day["open_month"], $special_day["open_day"]),
```
Comment (chrismiceli) [inline api/v1_projects.inc:364]: do we ever care about time zones or are these days always UTC and no concern with time zones in the "future"

File: api/dp-openapi.yaml [hunk 3]
```diff
@@ -189,6 +249,36 @@ paths:
...
+      parameters:
+      - name: enabled
+        in: query
+        description: Only return enabled special days
+        required: false
+        schema:
+          type: boolean
+      responses:
+        200:
+          description: List of valid character suites
```
[the remark this answers is not in the document]
Comment (cpeel) [reply api/dp-openapi.yaml:266]: Fixed it - thanks for catching that copy/pasta error.

**Review by chrismiceli — APPROVED** [review]

File: api/v1.inc (hunk 1 above)
Comment (cpeel) [reply api/v1.inc:28]: The ApiRouter evaluates static routes over parameter-based routes so the order isn't important. Although to be fair I had to look back at the code to confirm that for certain.

File: api/v1_projects.inc (hunk 2 above)
Comment (cpeel) [reply api/v1_projects.inc:364]: DP time is all "server time", so it isn't UTC and we also don't need to worry about timezones. It's an interesting artifact from the initial developers.

Comment (srjfoo) [reply api/v1_projects.inc:364]: See the comment [here](https://github.com/DistributedProofreaders/dproofreaders/blob/d01225fed2fddcad30ab069a46d0a9d355293175/crontab/onoff_special_event_queues.php#L24). That applies only to defined special days, though, not birthdays and otherdays.

Today's been busier than I expected -- I haven't done any reviewing or testing, yet, but wanted to go ahead and point to the time zone info. I don't know if that's relevant to the API, though.

File: api/dp-openapi.yaml [hunk 4]
```diff
@@ -626,6 +751,33 @@ components:
...
+        comment:
+          type: string
+        url:
+          type: string
+          format: url
+        color:
+          type: string
+        symbol:
+          type: string
+        date_open:
```
Comment (70ray) [inline api/dp-openapi.yaml:772]: Why not keep month and date as separate fields? then the user could more easily format the date however they wanted to.

Comment (cpeel) [reply api/dp-openapi.yaml:772]: Yeah, that's probably better. How does `date_open_month` and `date_open_day` sound as field names?

Comment (70ray) [reply api/dp-openapi.yaml:772]: Sounds alright to me.

Comment (srjfoo) [reply api/dp-openapi.yaml:772]: Works for me, too.

File: api/v1_projects.inc [hunk 5]
```diff
@@ -316,3 +317,86 @@ function api_v1_projects_pagerounds($method, $data, $query_params)
...
+            continue;
+        }
+
+        $return_data[] = [
+            "id" => $image_source["code_name"],
+            "name" => $image_source["display_name"],
+            "name_full" => $image_source["full_name"],
+            "url" => $image_source["url"],
+            "credit" => $image_source["credit"],
+            "enabled" => $image_source["is_active"] == 1,
```
[the remark this answers is not in the document]
Comment (cpeel) [reply api/v1_projects.inc:397]: I don't think there is, no. An image source is only usable / enabled if it's not pending and disabled which is what this is representing. The primary use of this API, at least initially, is going to be getting a list of special days to use within a project create / update call and those can only be used if they are enabled.

**Review by srjfoo — APPROVED** [review]

File: api/dp-openapi.yaml (hunk 4 above)
Comment (cpeel) [reply api/dp-openapi.yaml]: Updated.

**Review by srjfoo — APPROVED** [review]

**Pull request merged**